### PR TITLE
feat: workspaces and projects [DET-6461]

### PR DESCRIPTION
**Pull request by EmilyBonar**
## Description
This PR merges in a lot of work that has been done on Workspaces on a feature branch. This includes, but is not limited to:

- removing the Dashboard and Experiment List pages
- adding the Uncategorized, Workspace List, Workspace Details, and Project Details pages
- adding pinned workspaces to the navigation bar
- adding workspace and project commands to the CLI

## Test Plan

Extensive manual testing has taken place, including a bug bash.

Current Test Cluster: not up right now

for credentials visit your friendly local slack #workspaces.

[Issues being tracked](https://determinedai.atlassian.net/browse/DET-7330): mark something as P1 if you think it needs to be addressed before merge.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: harness/determined/cli/workspace.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+
+@authentication.required
+def list_workspace_projects(args: Namespace) -> None:
+    sess = setup_session(args)
+    w = workspace_by_name(sess, args.workspace_name)
+
+    orderArg = bindings.v1OrderBy[f"ORDER_BY_{args.order_by.upper()}"]
+    sortArg = bindings.v1GetWorkspaceProjectsRequestSortBy[f"SORT_BY_{args.sort_by.upper()}"]
+    projects = bindings.get_GetWorkspaceProjects(
```
Comment (ioga) [inline harness/determined/cli/workspace.py:57]: pagination?

File: harness/determined/cli/project.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    render.tabulate_or_csv(PROJECT_HEADERS, [values], False)
+
+
+def project_by_name(
+    sess: session.Session, workspace_name: str, project_name: str
+) -> Tuple[v1Workspace, v1Project]:
+    w = workspace_by_name(sess, workspace_name)
+    p = bindings.get_GetWorkspaceProjects(sess, id=w.id, name=project_name).projects
+    if len(p) == 0:
+        raise Exception(f'No project found on this workspace with name: "{project_name}"')
```
[the remark this answers is not in the document]
Comment (mapmeld) [reply harness/determined/cli/project.py:67]: ✅

File: master/internal/db/postgres_cluster.go [hunk 3]
```diff
@@ -51,6 +51,24 @@ SELECT jsonb_build_object(
 	'num_allocations', (SELECT count(*) FROM allocations),
     'num_allocations_today', (
 		SELECT count(*) FROM allocations WHERE DATE(start_time) >= CURRENT_DATE
+	),
+	'num_workspaces', (SELECT count(*) FROM workspaces),
+	'num_projects', (SELECT count(*) FROM projects),
+	'avg_projects_per_workspace', (
+		(SELECT count(*) FROM projects WHERE id > 1)::float
```
[the remark this answers is not in the document]
Comment (mapmeld) [reply master/internal/db/postgres_cluster.go:58]: The goal is to get a decimal / percentage here instead of Postgres's int / int = int behavior (i.e. `SELECT 3/4;` returns 0)

File: master/internal/db/postgres_experiments.go [hunk 4]
```diff
@@ -26,16 +28,37 @@ const (
 	min  = "min"
 )
 
+// ProjectFromNames returns a project's ID if it exists in the given workspace.
+func (db *PgDB) ProjectFromNames(workspaceName string, projectName string) (int, error) {
```
[the remark this answers is not in the document]
Comment (mapmeld) [reply master/internal/db/postgres_experiments.go:32]: ✅

File: harness/determined/cli/project.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mapmeld) [reply harness/determined/cli/project.py:67]: ok I had said checkmark but actually the api.errors.NotFoundException is based on a `requests.Response` not an error string. Generating a Response object is possible here but not a good fit - as this is CLI code running locally. We do a good job of returning an appropriate not-found server error in the Go code

```
	case db.ErrNotFound:
		return nil, status.Errorf(
			codes.NotFound, "project (%d) not found", id)
```

File: master/internal/api_project.go [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+	case db.ErrNotFound:
+		return nil, status.Errorf(
+			codes.NotFound, "project (%d) not found", id)
+	default:
+		return p, errors.Wrapf(err,
+			"error fetching project (%d) from database", id)
+	}
+}
+
+func (a *apiServer) ConfirmParentWorkspaceUnarchived(pid int32) error {
```
[the remark this answers is not in the document]
Comment (mapmeld) [reply master/internal/api_project.go:32]: ✅

File: master/internal/api_workspace.go [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/proto/pkg/apiv1"
+	"github.com/determined-ai/determined/proto/pkg/workspacev1"
+)
+
+func (a *apiServer) GetWorkspaceFromID(id int32, userID int32) (*workspacev1.Workspace,
```
[the remark this answers is not in the document]
Comment (mapmeld) [reply master/internal/api_workspace.go:19]: ✅

File: harness/determined/cli/workspace.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+
+
+@authentication.required
+def list_workspaces(args: Namespace) -> None:
+    # limit, name, offset, users
+    orderArg = bindings.v1OrderBy[f"ORDER_BY_{args.order_by.upper()}"]
+    sortArg = bindings.v1GetWorkspacesRequestSortBy[f"SORT_BY_{args.sort_by.upper()}"]
+    workspaces = bindings.get_GetWorkspaces(
+        setup_session(args), orderBy=orderArg, sortBy=sortArg
+    ).workspaces
```
[the remark this answers is not in the document]
Comment (mapmeld) [reply harness/determined/cli/workspace.py]: added in pagination commit

File: harness/determined/cli/workspace.py (hunk 1 above)
Comment (mapmeld) [reply harness/determined/cli/workspace.py]: added in pagination commit

File: harness/determined/cli/project.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+    (w, p) = project_by_name(sess, args.workspace_name, args.project_name)
+    kwargs: Dict[str, Any] = {
+        "id": p.id,
+        "orderBy": bindings.v1OrderBy[f"ORDER_BY_{args.order_by.upper()}"],
+        "sortBy": bindings.v1GetProjectExperimentsRequestSortBy[f"SORT_BY_{args.sort_by.upper()}"],
+    }
+    if not args.all:
+        kwargs["users"] = [authentication.must_cli_auth().get_session_user()]
+        kwargs["archived"] = "false"
+    experiments = bindings.get_GetProjectExperiments(sess, **kwargs).experiments
```
[the remark this answers is not in the document]
Comment (mapmeld) [reply harness/determined/cli/project.py]: added in pagination commit

File: master/static/migrations/20220422101847_pin-workspaces.up.sql [hunk 9]
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE workspace_pins (
+  id SERIAL PRIMARY KEY,
+  workspace_id INT REFERENCES workspaces(id),
```
[the remark this answers is not in the document]
Comment (mapmeld) [reply master/static/migrations/20220422101847_pin-workspaces.up.sql:3]: 👍

File: master/static/srv/delete_project.sql [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  AND NOT immutable
+  AND (user_id = $2 OR $3 IS TRUE
+    OR workspace_id IN (SELECT id FROM owned_workspaces)
+  )
+),
+exper AS (
+  UPDATE experiments
+  SET project_id = 1
+  WHERE project_id IN (SELECT id FROM proj)
+)
```
[the remark this answers is not in the document]
Comment (mapmeld) [reply master/static/srv/delete_project.sql:18]: Here's my proposed fix (I think I sent an event invite where we can iterate / discuss any remaining issues) https://github.com/determined-ai/determined/pull/4224

File: harness/determined/cli/project.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ioga) [reply harness/determined/cli/project.py:67]: ok, then maybe add a new `class EmptyResultException(BadResponseException):`?

**Review by mapmeld — APPROVED** [review]
I'm going to approve, though I have more familiarity with the non-WebUI code changes. I think we should confirm with @ioga that we addressed Backend issues, and see if WebUI looks right.

File: e2e_tests/tests/cluster/test_workspace_org.py [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+from tests import config as conf
+from tests.experiment import run_basic_test, wait_for_experiment_state
+
+
+@pytest.mark.e2e_cpu
+def test_workspace_org() -> None:
+    master_url = conf.make_master_url()
+    certs.cli_cert = certs.default_load(master_url)
+    authentication.cli_auth = authentication.Authentication(master_url, try_reauth=True)
+    sess = session.Session(master_url, "determined", authentication.cli_auth, certs.cli_cert)
```
[the remark this answers is not in the document]
Comment (ioga) [reply e2e_tests/tests/cluster/test_workspace_org.py:16]: any comment on this?..

File: master/static/srv/pin_workspace.sql [hunk 12]
```diff
@@ -0,0 +1,3 @@
+INSERT INTO workspace_pins (workspace_id, user_id, created_at)
+VALUES ($1, $2, NOW())
```
[the remark this answers is not in the document]
Comment (ioga) [reply master/static/srv/pin_workspace.sql:2]: any comment on this?

File: harness/determined/cli/project.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ioga) [reply harness/determined/cli/project.py:67]: I'd still prefer a typed exception hierarchy.

File: master/internal/api_project.go [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+		apiv1.GetProjectExperimentsRequest_SORT_BY_NAME:          "name",
+		apiv1.GetProjectExperimentsRequest_SORT_BY_START_TIME:    "start_time",
+		apiv1.GetProjectExperimentsRequest_SORT_BY_END_TIME:      "end_time",
+		apiv1.GetProjectExperimentsRequest_SORT_BY_STATE:         "state",
+		apiv1.GetProjectExperimentsRequest_SORT_BY_NUM_TRIALS:    "num_trials",
+		apiv1.GetProjectExperimentsRequest_SORT_BY_PROGRESS:      "COALESCE(progress, 0)",
+		apiv1.GetProjectExperimentsRequest_SORT_BY_USER:          "username",
+		apiv1.GetProjectExperimentsRequest_SORT_BY_FORKED_FROM:   "forked_from",
+		apiv1.GetProjectExperimentsRequest_SORT_BY_RESOURCE_POOL: "resource_pool",
+	}
```
[the remark this answers is not in the document]
Comment (ioga) [reply master/internal/api_project.go:93]: thoughts?

File: master/internal/api_project.go [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+		return nil, err
+	}
+	if w.Immutable {
+		return nil, errors.Errorf("workspace (%v) is immutable and cannot add new projects.",
+			w.Id)
+	}
+	if w.Archived {
+		return nil, errors.Errorf("workspace (%v) is archived and cannot add new projects.",
+			w.Id)
+	}
```
[the remark this answers is not in the document]
Comment (ioga) [reply master/internal/api_project.go:263]: thoughts?

File: master/internal/db/postgres_experiments.go [hunk 15]
```diff
@@ -26,16 +28,37 @@ const (
...
+	if err != nil {
+		return 1, err
+	}
+	p := projectv1.Project{}
+	err = db.Query("get_project_from_name", &p, w.Id, projectName)
+	if err != nil {
+		return 1, err
+	}
+	if p.Id < 1 {
+		return 1, fmt.Errorf("workspace and project names did not match any known project")
```
[the remark this answers is not in the document]
Comment (ioga) [reply master/internal/db/postgres_experiments.go:44]: thoughts?

**Comment by ioga** [comment]
I think we've resolved the major issues.
I've bumped a few of my earlier comments that went unaddressed, most of them are on code style or quality.

File: master/internal/api_project.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (mapmeld) [reply master/internal/api_project.go:263]: ✅ added an argument `rejectImmutable` to `GetWorkspaceByID`, implementing on api_project and api_workspace

File: master/internal/db/postgres_experiments.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (mapmeld) [reply master/internal/db/postgres_experiments.go]: fixing with `return 1, ErrNotFound`

File: master/internal/api_project.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (mapmeld) [reply master/internal/api_project.go:93]: moved to shared `SortExperimentBy` type

File: harness/determined/cli/project.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mapmeld) [reply harness/determined/cli/project.py:67]: Should I be handling CLI exceptions more like we do in Experiments?
```
    except Exception as e:
        print("Failed to initialize git repository at ", "{}: {}".format(repo_path, e))
        sys.exit(1)
```

[the remark this answers is not in the document]
Comment (ioga) [reply harness/determined/cli/project.py:67]: I'd say throwing an exception should be enough. calling `sys.exit` in the middle of the code is a bad pattern in my opinion, because you can't effectively reuse these functions elsewhere then.

[the remark this answers is not in the document]
Comment (mapmeld) [reply harness/determined/cli/project.py]: ok here's my take on `EmptyResultException` and `ForbiddenException` in the areas where we are having this part of the CLI raise exceptions.  If this looks good, I'll make a ticket to apply these to the Models CLI at least.

[the remark this answers is not in the document]
Comment (mapmeld) [reply harness/determined/cli/project.py]: Link: https://github.com/determined-ai/determined/pull/4203/commits/8258a165e5e2dbf053745ce93f0796dc62f60bf5

File: webui/react/src/utils/experiment.ts [hunk 16]
```diff
@@ -74,3 +96,115 @@ export const upgradeConfig = (config: RawJson): RawJson => {
...
+
+  [ExperimentAction.Archive]: (experiment, user) =>
+    !experiment.parentArchived && !experiment.archived && terminalRunStates.has(experiment.state),
+
+  [ExperimentAction.Cancel]: (experiment, user) =>
+    cancellableRunStates.has(experiment.state),
+
+  [ExperimentAction.CompareTrials]: alwaysTrueExperimentChecker,
+
+  [ExperimentAction.ContinueTrial]: isExperimentModifiable,
```
Comment (trentwatt) [inline webui/react/src/utils/experiment.ts:132]: should also check `len(trials) == 1`

File: webui/react/src/components/BreadcrumbBar.tsx [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+
+interface Props {
+  experiment?: ExperimentBase;
+  extra?: React.ReactNode;
+  id: number;
+  project?: Project;
+  trial?: TrialDetails;
+  type: 'project' | 'experiment' | 'trial';
+  workspace?: Workspace;
+}
```
Comment (hkang1) [inline webui/react/src/components/BreadcrumbBar.tsx:24]: **question:** So this is a "smart" component where it's doing a lot of API calls given types and entities and then figuring out how to render after collecting all the info. Curious if there is a strong reason to approach it this way. One possibility is that we want to centralize the logic of fetching data for filling out the breadcrumb component.

Alternatives could be to have a context that keeps track of traversed entities, and at each traversal into specifics, to save the context of that particular item. (possibly in `Store` as something like `breadcrumbs`). For example, when traversing into an experiment detail, `breadcrumbs` would be `[ <experiment> ]`. Then diving into one of the trials, would be something like `breadcrumbs = [ <experiment x>, <trial x> ]`, etc. BreadcrumbBar can just consume the store directly, with the goal of avoid having to (potentially extra) API calls.

File: webui/react/src/omnibar/tree-extension/trees/goto.ts [hunk 18]
```diff
@@ -59,10 +55,6 @@ const locations: TreeNode[] = [
     },
     title: 'trial',
   },
-  {
-    onAction: visitAction(paths.dashboard()),
-    title: 'dashboard',
-  },
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply webui/react/src/omnibar/tree-extension/trees/goto.ts:65]: https://determinedai.atlassian.net/browse/DET-5382 I added a note

File: webui/react/src/components/Breadcrumb.stories.tsx [hunk 19]
```diff
@@ -15,7 +15,7 @@ export const Default = (): React.ReactNode => (
     <Breadcrumb.Item>
       <Space align="center" size="small">
         <Icon name="experiment" size="small" />
-        <Link path={paths.experimentList()}>Experiments</Link>
+        {/* <Link path={paths.experimentList()}>Experiments</Link> */}
```
[the remark this answers is not in the document]
Comment (EmilyBonar) [reply webui/react/src/components/Breadcrumb.stories.tsx:18]: This file really just needs to be removed tbh.

File: webui/react/src/components/ActionSheet.module.scss [hunk 20]
```diff
@@ -13,16 +13,26 @@
     bottom: 0;
     display: flex;
     flex-direction: column;
+    max-height: 90%;
```
[the remark this answers is not in the document]
Comment (EmilyBonar) [reply webui/react/src/components/ActionSheet.module.scss:16]: The ActionSheet comes up from the bottom of the screen on mobile and is used for the list of pinned workspaces (and also the overflow menu). This makes it so if you pinned a large number of workspaces, the list will start scrolling without covering the whole screen, like this:
Screen Shot 2022-06-10 at 13 55 37

File: webui/react/src/pages/Cluster/ClusterOverview.tsx [hunk 21]
```diff
@@ -86,7 +86,7 @@ const ClusterOverview: React.FC = () => {
     setTotal(resourcePools.length || 0);
   }, [ resourcePools ]);
 
-  usePolling(fetchResourcePools, { interval: 10000 });
+  usePolling(fetchResourcePools, { interval: 10000, rerunOnNewFn: true });
```
[the remark this answers is not in the document]
Comment (EmilyBonar) [reply webui/react/src/pages/Cluster/ClusterOverview.tsx:89]: Makes sense, though the Cluster page is no longer in use.

File: webui/react/src/pages/Clusters/ClustersOverview.tsx [hunk 22]
```diff
@@ -29,7 +29,7 @@ const ClusterOverview: React.FC = () => {
   const fetchAgents = useFetchAgents(canceler);
   const fetchResourcePools = useFetchResourcePools(canceler);
 
-  usePolling(fetchResourcePools, { interval: 10000 });
+  usePolling(fetchResourcePools, { interval: 10000, rerunOnNewFn: true });
```
[the remark this answers is not in the document]
Comment (EmilyBonar) [reply webui/react/src/pages/Clusters/ClustersOverview.tsx:32]: Makes sense, though the Cluster page is no longer in use.

File: webui/react/src/components/WorkspaceIcon.module.scss [hunk 23]
```diff
@@ -0,0 +1,11 @@
+.base {
+  aspect-ratio: 1;
+  background-color: var(--theme-colors-monochrome-15);
+  border-radius: 8px;
```
[the remark this answers is not in the document]
Comment (EmilyBonar) [reply webui/react/src/components/WorkspaceIcon.module.scss:4]: 👍

File: webui/react/src/components/BreadcrumbBar.tsx (hunk 17 above)
Comment (EmilyBonar) [reply webui/react/src/components/BreadcrumbBar.tsx:24]: Yeah, I wanted to have all of the logic for constructing the breadcrumbs in one place, and not have to worry about different pages handling it differently. With putting data into the Store, I worry about that somehow getting into a bad state and showing incorrect/stale data. Open to hearing other ideas though.

File: webui/react/src/pages/ProjectDetails.tsx [hunk 24]
```diff
@@ -1,67 +1,78 @@
...
+  getActionsForExperimentsUnion,
+  getProjectExperimentForExperimentItem,
+} from 'utils/experiment';
 import { alphaNumericSorter } from 'utils/sort';
-import { isTaskKillable, taskFromExperiment } from 'utils/task';
 import { getDisplayName } from 'utils/user';
 import { openCommand } from 'wait';
 
 import { RecordKey } from '../shared/types';
 import { ErrorLevel } from '../shared/utils/error';
```
Comment (hkang1) [inline webui/react/src/pages/ProjectDetails.tsx:67]: **nit:** don't have to, but convenient to just have `shared/...` instead of `../shared/...`

**Review by hkang1 — COMMENTED** [review]
Another round of review (143 files out of 180) will get to the remainder ASAP

File: webui/react/src/components/NotesCard.tsx [hunk 25]
```diff
@@ -65,26 +99,40 @@ const NotesCard: React.FC<Props> = ({ disabled = false, notes, onSave, style }:
...
+        />
+      )}>
       <Spinner spinning={isLoading}>
         <Markdown
           editing={isEditing}
           markdown={isEditing ? editedNotes : notes}
-          onChange={setEditedNotes}
-          onClick={() => { if (notes === '') editNotes(); }}
+          onChange={handleEditedNotes}
+          onClick={(e: React.MouseEvent) => { if (e.detail > 1 || notes === '') editNotes(); }}
```
Comment (hkang1) [inline webui/react/src/components/NotesCard.tsx:126]: **nit suggestion:** use same pattern as `handleEditedNotes` (via useCallback)?

File: webui/react/src/components/NotesCard.tsx [hunk 26]
```diff
@@ -65,26 +99,40 @@ const NotesCard: React.FC<Props> = ({ disabled = false, notes, onSave, style }:
...
         />
       </Spinner>
       <Prompt
         message={(newLocation) => {
+          const isSameExperiment =
+          location.pathname.split('/')[0] === 'experiment' &&
+            newLocation.pathname.startsWith(location.pathname.split('/').slice(0, -1).join('/'));
```
Comment (hkang1) [inline webui/react/src/components/NotesCard.tsx:133]: **suggestion:** we'll probably use this again in the future for other areas of the app. Make it into a routes utility function?

File: webui/react/src/components/PaginatedNotesCard.module.scss [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+        width: 100%;
+
+        & > span {
+          flex-grow: 1;
+        }
+        .action {
+          visibility: hidden;
+        }
+        &:hover {
+          background-color: var(--theme-colors-monochrome-12);
```
Comment (hkang1) [inline webui/react/src/components/PaginatedNotesCard.module.scss:37]: **chore:** not sure what this one should be, `surface`?
```
background-color: var(--theme-surface-strong);
```

File: webui/react/src/components/PaginatedNotesCard.module.scss [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+    align-items: center;
+    display: flex;
+    flex-direction: column;
+    gap: 18px;
+
+    & > * {
+      margin: 0;
+    }
+    & > i {
+      color: var(--theme-colors-monochrome-8)
```
Comment (hkang1) [inline webui/react/src/components/PaginatedNotesCard.module.scss:74]: **chore:** theme update (also not sure what works here, may need to try stuff out)

File: webui/react/src/components/PaginatedNotesCard.module.scss [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+    }
+  }
+}
+
+@media only screen and (max-width: $breakpoint-tablet) {
+  .base {
+    flex-direction: column-reverse;
+
+    .notesContainer {
+      min-height: 90%;
```
Comment (hkang1) [inline webui/react/src/components/PaginatedNotesCard.module.scss:84]: **suggestion:** if the 10% is to make room for chrome or nav components, a percentage will be pretty brittle in terms of responsive design. A more fool proof method could be the use of `calc()` function to calculate exactly what's needed dynamically.

File: webui/react/src/components/PaginatedNotesCard.tsx [hunk 30]
```diff
@@ -0,0 +1,203 @@
...
+
+import NotesCard from './NotesCard';
+import css from './PaginatedNotesCard.module.scss';
+
+interface Props {
+  disabled?: boolean;
+  notes: Note[];
+  onDelete: (pageNumber: number) => void;
+  onNewPage: () => void;
+  onSave: (notes: Note[]) => void;
```
Comment (hkang1) [inline webui/react/src/components/PaginatedNotesCard.tsx:17]: **question:** below there are call patterns of `await onSave(...)`. Maybe this needs to be `(notes: Note[]) => Promise<void>`?

File: webui/react/src/components/PaginatedNotesCard.tsx [hunk 31]
```diff
@@ -0,0 +1,203 @@
...
+      <div className={css.notesContainer}>
+        <NotesCard
+          disabled={disabled}
+          extra={(
+            <Dropdown
+              overlay={() => ActionMenu(currentPage)}
+              trigger={[ 'click' ]}>
+              <div style={{ cursor: 'pointer' }}>
+                <Icon name="overflow-horizontal" />
+              </div>
```
Comment (hkang1) [inline webui/react/src/components/PaginatedNotesCard.tsx:186]: **nit:** use className on `Icon` directly to shorten this to one line? For example:
```
<Icon className={css.pointer} name="overflow-horizontal" />
```

File: webui/react/src/components/PaginatedNotesCard.tsx [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+                      'var(--theme-colors-monochrome-12)' :
+                      undefined,
+                  }}
+                  onClick={() => handleSwitchPage(idx)}>
+                  <span>{note.name}</span>
+                  {!disabled && (
+                    <Dropdown
+                      overlay={() => ActionMenu(idx)}
+                      trigger={[ 'click' ]}>
+                      <div className={css.action} onClick={e => e.stopPropagation()}>
```
Comment (hkang1) [inline webui/react/src/components/PaginatedNotesCard.tsx:166]: **suggestion:** the stopPropagation is used pretty frequently in the context of `ActionDropdown`. Could be a useful export from `ActionDropdown.tsx` and have it defined once there.

File: webui/react/src/components/PaginatedNotesCard.tsx [hunk 33]
```diff
@@ -0,0 +1,203 @@
...
+                key={idx}
+                overlay={() => ActionMenu(idx)}
+                trigger={[ 'contextMenu' ]}>
+                <li
+                  className={css.listItem}
+                  style={{
+                    borderColor: idx === currentPage ?
+                      'var(--theme-colors-monochrome-12)' :
+                      undefined,
+                  }}
```
Comment (hkang1) [inline webui/react/src/components/PaginatedNotesCard.tsx:159]: **suggestion:** apply it as a dynamic class instead of inline style (e.g. `css.listItem` and `css.listItem css.currentListItem` or something like that)

File: webui/react/src/components/PaginatedNotesCard.tsx [hunk 34]
```diff
@@ -0,0 +1,203 @@
...
+
+  useEffect(() => {
+    if (notes.length === 0) return;
+    if (currentPage < 0) { setCurrentPage(0); fireNoteChangeSignal(); }
+    if (currentPage >= notes.length) { setCurrentPage(notes.length - 1); fireNoteChangeSignal(); }
+  }, [ currentPage, notes.length, fireNoteChangeSignal ]);
+
+  useEffect(() => {
+    setEditedContents(notes?.[currentPage]?.contents ?? '');
+    setEditedName(notes?.[currentPage]?.name ?? '');
```
Comment (hkang1) [inline webui/react/src/components/PaginatedNotesCard.tsx:114]: **question:** would it be better to not make any changes if the input is invalid? for example:
```
setEditedContents(prev => notes?.[currentPage]?.contents ?? prev);
```

**Review by hkang1 — APPROVED** [review]
ok I think I got through all the Web changes. Great work 🍾 🎉 ! I provided a lot of feedback throughout the PR. Lot's of areas of minor improvements, but nothing was blocking. If you agree with the improvements, but don't feel it needs to make it into this PR, please create tickets and with the improvement so we can address them in the near future.

File: webui/react/src/pages/WorkspaceDetails/WorkspaceDetailsHeader.tsx [hunk 35]
```diff
@@ -0,0 +1,89 @@
...
+                 workspace.archived
+                || (!curUser?.isAdmin && curUser?.username !== workspace.username)}
+              maxLength={80}
+              value={workspace.name}
+              onSave={handleNameChange}
+            />
+          </h1>
+          {workspace.archived && (
+            <Tooltip title="Archived">
+              <div>
```
[the remark this answers is not in the document]
Comment (EmilyBonar) [reply webui/react/src/pages/WorkspaceDetails/WorkspaceDetailsHeader.tsx:62]: If you don't put a `div` (or some other element) there, the icon doesn't show up. I believe this is because the Tooltip attaches to that element.

File: webui/react/src/pages/WorkspaceDetails/ProjectCard.tsx [hunk 36]
```diff
@@ -0,0 +1,89 @@
...
+import { paths } from 'routes/utils';
+import Icon from 'shared/components/Icon/Icon';
+import { routeToReactUrl } from 'shared/utils/routes';
+import { DetailedUser, Project } from 'types';
+
+import ProjectActionDropdown from './ProjectActionDropdown';
+import css from './ProjectCard.module.scss';
+
+interface Props {
+  curUser?: DetailedUser;
```
[the remark this answers is not in the document]
Comment (EmilyBonar) [reply webui/react/src/pages/WorkspaceDetails/ProjectCard.tsx:16]: True, but we use `curUser` all over our codebase, so this is for consistency

File: webui/react/src/pages/WorkspaceDetails/ProjectActionDropdown.tsx [hunk 37]
```diff
@@ -0,0 +1,141 @@
...
+    return null;
+  }
+
+  return children ? (
+    <Dropdown
+      disabled={menuItems.length === 0}
+      overlay={(
+        <Menu>
+          {menuItems}
+        </Menu>
```
[the remark this answers is not in the document]
Comment (EmilyBonar) [reply webui/react/src/pages/WorkspaceDetails/ProjectActionDropdown.tsx:112]: That would require refactoring menuItems to be a `MenuItem[]` instead of a `React.ReactNode[]`. This is harder than it sounds.

File: webui/react/src/components/PaginatedNotesCard.tsx (hunk 31 above)
Comment (EmilyBonar) [reply webui/react/src/components/PaginatedNotesCard.tsx:186]: Similar to the tooltip previously, the `div` is necessary for the Dropdown to attach to.
